packages/datadog-shimmer/src/shimmer.js

```diff
@@ -1,85 +1,88 @@
 'use strict'
 
+const skipMethods = new Set([
+  'caller',
+  'arguments',
+  'name',
+  'length'
+])
+
 function copyProperties (original, wrapped) {
-  // TODO getPrototypeOf is not fast. Should we instead do this in specific
-  // instrumentations where needed?
-  const proto = Object.getPrototypeOf(original)
-  if (proto !== Function.prototype) {
+  if (original.constructor !== wrapped.constructor) {
+    const proto = Object.getPrototypeOf(original)
     Object.setPrototypeOf(wrapped, proto)
   }
-  const props = Object.getOwnPropertyDescriptors(original)
-  const keys = Reflect.ownKeys(props)
-
-  for (const key of keys) {
-    try {
-      Object.defineProperty(wrapped, key, props[key])
-    } catch (e) {
-      // TODO: figure out how to handle this without a try/catch
+
+  const ownKeys = Reflect.ownKeys(original)
+  if (original.length !== wrapped.length) {
+    Object.defineProperty(wrapped, 'length', { value: original.length, configurable: true })
+  }
+  if (original.name !== wrapped.name) {
+    Object.defineProperty(wrapped, 'name', { value: original.name, configurable: true })
+  }
+  if (ownKeys.length !== 2) {
+    for (const key of ownKeys) {
+      if (skipMethods.has(key)) continue
+      const descriptor = Object.getOwnPropertyDescriptor(original, key)
+      if (descriptor.writable || descriptor.configurable || !Object.hasOwn(wrapped, key)) {
+        Object.defineProperty(wrapped, key, descriptor)
+      }
     }
   }
 }
 
 function wrapFunction (original, wrapper) {
-  if (typeof original === 'function') assertNotClass(original)
-
   const wrapped = wrapper(original)
 
-  if (typeof original === 'function') copyProperties(original, wrapped)
+  if (typeof original === 'function') {
+    assertNotClass(original)
```

Comment:
This is a slight change in behaviour, because the wrapper function still gets called even if it's a class, whereas before it wouldn't (it would throw).

In practice, this shouldn't be a problem because we'll throw either way, and any side effects of wrapping likely won't matter much since we'll be aborting instrumentation anyway.

This is probably fine, since tests are currently passing, and functions don't _typically_ become classes in new versions anymore (this used to be common when ES5 syntax was new, but it's been a decade now 🤣).

Non-blocker, just notable.

Reply:
The reasoning for the change is: the safe guard is there for our tests, so that we detect this early on. It would definitely fail the same as before, even if a class is passed through, just after passing the original to the wrapper. That could theoretically fail earlier but since both cases end up with an error, the behavior does not really change.